In [ ]:
from zipfile import ZipFile
zip_ref = ZipFile('/content/zara_clothes_dataset.zip','r')
zip_ref.extractall('/content')
zip_ref.close()

In [ ]:
men_dataset_path = "/content/Men/Men"
men_dataset_path

'/content/Men/Men'

In [ ]:
import os
import pandas as pd
men_df = []
for i in os.listdir(men_dataset_path):
  men_df.append(i)
  print(i)

men_data = pd.concat([pd.read_csv(os.path.join(men_dataset_path,f)) for f in men_df])
men_data
#men_data.to_csv("men_data.csv")

ZARA ATHLETICZ.csv
PERFUMES.csv
BAGS_BACKPACKS.csv
SHORTS.csv
TROUSERS.csv
SWIMWEAR.csv
LINEN.csv
JACKETS.csv
SHIRTS.csv
SUITS.csv
JEANS.csv
BLAZERS.csv
OVERSHIRTS.csv
SWEATERS_CARDIGANS.csv
ACCESSORIES.csv
ZARA ORIGINS.csv
SHOES.csv
POLO SHIRTS.csv
HOODIES_SWEATSHIRTS.csv
T-SHIRTS.csv
BEAUTY.csv


,Unnamed: 0,product_name,link,product_images,price,details
0,0,TECHNICAL TRAINING TROUSERS,https://www.zara.com/in/en/technical-training-...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of lightweight and stretchy tech...
1,1,BASIC TRAINING SHORTS,https://www.zara.com/in/en/basic-training-shor...,[{'https://static.zara.net/photos///2023/I/0/1...,"₹ 2,290.00",Workout shorts made of lightweight stretchy te...
2,2,COTTON HOODIE,https://www.zara.com/in/en/cotton-hoodie-p0076...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,990.00",Hoodie made of compact cotton fabric. Adjustab...
3,3,JOGGER WAIST COTTON TROUSERS,https://www.zara.com/in/en/jogger-waist-cotton...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of compact cotton fabric. Adjust...
4,6,T-SHIRT WITH PRINTED LOGO,https://www.zara.com/in/en/t-shirt-with-printe...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 2,290.00",T-shirt made of a cotton fabric. Round neck an...
...,...,...,...,...,...,...
12,34,FACE BRUSH #2,https://www.zara.com/in/en/face-brush--2-p2414...,[],"₹ 1,290.00",Designed to apply powder and cream products. P...
13,36,KABUKI BRUSH,https://www.zara.com/in/en/kabuki-brush-p24140...,[{'https://static.zara.net/photos///2022/V/2/1...,"₹ 1,290.00","Short and smooth dense-bristle brush, ideal fo..."
14,40,EYE BRUSH #2,https://www.zara.com/in/en/eye-brush--2-p24140...,[{'https://static.zara.net/photos///2022/I/2/1...,₹ 590.00,Compact and flat eye brush specially designed ...
15,41,EYE BRUSH #1,https://www.zara.com/in/en/eye-brush--1-p24140...,[{'https://static.zara.net/photos///2022/I/2/1...,₹ 590.00,Small eye brush designed to blend or apply eye...


In [ ]:
men_data.shape

(1478, 6)

In [ ]:
men_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1478 entries, 0 to 16
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Unnamed: 0       1478 non-null   int64 
 1   product_name     1476 non-null   object
 2   link             1478 non-null   object
 3    product_images  1478 non-null   object
 4   price            1476 non-null   object
 5   details          1475 non-null   object
dtypes: int64(1), object(5)
memory usage: 80.8+ KB


In [ ]:
men_data.describe()

,Unnamed: 0
count,1478.000000
mean,138.661705
std,115.469314
min,0.000000
25%,42.250000
50%,109.000000
75%,214.000000
max,482.000000


In [ ]:
men_data.isnull().sum()

,0
Unnamed: 0,0
product_name,2
link,0
product_images,0
price,2
details,3


In [ ]:
men_data.drop("Unnamed: 0",axis=1,inplace=True)

In [ ]:
men_data['product_name'].value_counts()

,count
product_name,
CONTRAST PRINTED T-SHIRT,11
100% LINEN SHIRT,9
SLOGAN PRINT T-SHIRT,9
GEOMETRIC PRINT SWIMMING TRUNKS,8
100% LINEN SUIT TROUSERS,7
...,...
ULTIMATTE MATTE LIQUID LIPSTICK,1
TERRY SWEATSHIRT,1
TERRY TANK TOP,1


In [ ]:
men_data['price'].value_counts()

,count
price,
"₹ 3,290.00",291
"₹ 4,990.00",172
"₹ 3,990.00",154
"₹ 2,990.00",153
"₹ 2,590.00",148
"₹ 2,290.00",99
"₹ 5,990.00",72
"₹ 2,890.00",60
"₹ 1,990.00",52


In [ ]:
import ast
def extract_first_image_url(image_str):
    try:
        # Safely evaluate the string into a Python list of dictionaries
        image_list = ast.literal_eval(image_str)
        if image_list and isinstance(image_list, list):
            # Iterate through the dictionaries in the list
            for img_dict in image_list:
                if isinstance(img_dict, dict):
                    # The URL key is the full URL string itself
                    for key in img_dict.keys():
                        if key.startswith('https://static.zara.net/photos///'):
                            return key
        return None
    except (ValueError, SyntaxError):
        # Handle cases where the string is not a valid literal
        return None
# Apply the function to the ' product_images' column
men_data['image_url'] = men_data[' product_images'].apply(extract_first_image_url)
men_data.head()

,Unnamed: 0,product_name,link,product_images,price,details,image_url
0,0,TECHNICAL TRAINING TROUSERS,https://www.zara.com/in/en/technical-training-...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of lightweight and stretchy tech...,https://static.zara.net/photos///2023/I/0/2/p/...
1,1,BASIC TRAINING SHORTS,https://www.zara.com/in/en/basic-training-shor...,[{'https://static.zara.net/photos///2023/I/0/1...,"₹ 2,290.00",Workout shorts made of lightweight stretchy te...,https://static.zara.net/photos///2023/I/0/1/p/...
2,2,COTTON HOODIE,https://www.zara.com/in/en/cotton-hoodie-p0076...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,990.00",Hoodie made of compact cotton fabric. Adjustab...,https://static.zara.net/photos///2023/I/0/2/p/...
3,3,JOGGER WAIST COTTON TROUSERS,https://www.zara.com/in/en/jogger-waist-cotton...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of compact cotton fabric. Adjust...,https://static.zara.net/photos///2023/I/0/2/p/...
4,6,T-SHIRT WITH PRINTED LOGO,https://www.zara.com/in/en/t-shirt-with-printe...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 2,290.00",T-shirt made of a cotton fabric. Round neck an...,https://static.zara.net/photos///2023/I/0/2/p/...


**Reasoning**:
Now that the product images have been preprocessed, the next step in building the product similarity search engine is to clean the 'product_name' column. This involves handling missing values, converting text to lowercase, and removing punctuation to prepare it for TF-IDF vectorization. First, I will fill any missing values in the 'product_name' column with an empty string.



In [ ]:
men_data['product_name'].fillna('', inplace=True)
print(men_data['product_name'].isnull().sum())


0


/tmp/ipykernel_349/702133974.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  men_data['product_name'].fillna('', inplace=True)


**Reasoning**:
The previous code generated a `FutureWarning` because `inplace=True` was used with chained assignment. To address this and ensure future compatibility, I will update the code to reassign the result of `fillna` back to the 'product_name' column, which is the recommended practice.



In [ ]:
men_data['product_name'] = men_data['product_name'].fillna('')
print(men_data['product_name'].isnull().sum())

0


**Reasoning**:
To further clean the 'product_name' column for TF-IDF vectorization, I will convert all text to lowercase to ensure consistency and prevent variations in capitalization from being treated as different product names.



In [ ]:
men_data['product_name'] = men_data['product_name'].str.lower()
men_data.head()

,Unnamed: 0,product_name,link,product_images,price,details,image_url
0,0,technical training trousers,https://www.zara.com/in/en/technical-training-...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of lightweight and stretchy tech...,https://static.zara.net/photos///2023/I/0/2/p/...
1,1,basic training shorts,https://www.zara.com/in/en/basic-training-shor...,[{'https://static.zara.net/photos///2023/I/0/1...,"₹ 2,290.00",Workout shorts made of lightweight stretchy te...,https://static.zara.net/photos///2023/I/0/1/p/...
2,2,cotton hoodie,https://www.zara.com/in/en/cotton-hoodie-p0076...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,990.00",Hoodie made of compact cotton fabric. Adjustab...,https://static.zara.net/photos///2023/I/0/2/p/...
3,3,jogger waist cotton trousers,https://www.zara.com/in/en/jogger-waist-cotton...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of compact cotton fabric. Adjust...,https://static.zara.net/photos///2023/I/0/2/p/...
4,6,t-shirt with printed logo,https://www.zara.com/in/en/t-shirt-with-printe...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 2,290.00",T-shirt made of a cotton fabric. Round neck an...,https://static.zara.net/photos///2023/I/0/2/p/...


**Reasoning**:
To further clean the 'product_name' column and prepare it for TF-IDF vectorization, I will remove punctuation from the text. This helps in treating words like 'T-shirt' and 'Tshirt' as the same, improving the consistency and effectiveness of the vectorization process.



In [ ]:
import re
men_data['product_name'] = men_data['product_name'].apply(lambda x: re.sub(r'[^\w\s]', '', x))
men_data.head()

,Unnamed: 0,product_name,link,product_images,price,details,image_url
0,0,technical training trousers,https://www.zara.com/in/en/technical-training-...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of lightweight and stretchy tech...,https://static.zara.net/photos///2023/I/0/2/p/...
1,1,basic training shorts,https://www.zara.com/in/en/basic-training-shor...,[{'https://static.zara.net/photos///2023/I/0/1...,"₹ 2,290.00",Workout shorts made of lightweight stretchy te...,https://static.zara.net/photos///2023/I/0/1/p/...
2,2,cotton hoodie,https://www.zara.com/in/en/cotton-hoodie-p0076...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,990.00",Hoodie made of compact cotton fabric. Adjustab...,https://static.zara.net/photos///2023/I/0/2/p/...
3,3,jogger waist cotton trousers,https://www.zara.com/in/en/jogger-waist-cotton...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 3,290.00",Trousers made of compact cotton fabric. Adjust...,https://static.zara.net/photos///2023/I/0/2/p/...
4,6,tshirt with printed logo,https://www.zara.com/in/en/t-shirt-with-printe...,[{'https://static.zara.net/photos///2023/I/0/2...,"₹ 2,290.00",T-shirt made of a cotton fabric. Round neck an...,https://static.zara.net/photos///2023/I/0/2/p/...


**Reasoning**:
Now that the 'product_name' column has been cleaned by handling missing values, converting to lowercase, and removing punctuation, the next step is to apply TF-IDF vectorization. This will transform the text data into numerical feature vectors, which is essential for calculating similarity.



In [ ]:
term_to_check = 'shirt'

# Check if the singular form of the query term exists in the TF-IDF vocabulary
if term_to_check in tfidf_vectorizer.vocabulary_:
    print(f"'{term_to_check}' is in the TF-IDF vocabulary.")
else:
    print(f"'{term_to_check}' is NOT in the TF-IDF vocabulary.")

# Additionally, let's check if any product names in the original (cleaned) data actually contain 'shirt'
products_with_shirt = men_data[men_data['product_name'].str.contains(term_to_check, na=False)]
print(f"\nNumber of products containing '{term_to_check}' in their name (after cleaning): {len(products_with_shirt)}")
if not products_with_shirt.empty:
    print("Sample product names containing 'shirt':")
    for name in products_with_shirt['product_name'].head(5):
        print(f"- {name}")


'shirt' is in the TF-IDF vocabulary.

Number of products containing 'shirt' in their name (after cleaning): 480
Sample product names containing 'shirt':
- tshirt with printed logo
- hatthanan print tshirt
- technical training sweatshirt
- contrast printed tshirt
- contrast printed tshirt


**Reasoning**:
Since the term 'shirt' is present in the TF-IDF vocabulary, unlike 'shirts', I will now re-run the similarity search using 'shirt' as the query. This should yield relevant results with non-zero similarity scores, demonstrating the correct functioning of the updated search function.



In [ ]:
query = "shirt"
results = product_similarity_search(query, N=5)
print(f"Top 5 similar products for query '{query}':")
for index, row in results.iterrows():
    print(f"Product: {row['product_name']}, Similarity: {row['similarity_score']:.2f}, Image URL: {row['image_url']}")

Top 5 similar products for query 'shirt':
Product: shirt, Similarity: 1.00, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4087/360/700/2/w/448/4087360700_1_1_1.jpg?ts=1689148214465
Product: shirt, Similarity: 1.00, Image URL: https://static.zara.net/photos///2023/I/0/2/p/3268/306/707/2/w/448/3268306707_1_1_1.jpg?ts=1692174425473
Product: textured shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/2092/300/250/2/w/448/2092300250_1_1_1.jpg?ts=1690536385550
Product: textured shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/3090/510/800/2/w/448/3090510800_1_1_1.jpg?ts=1688652389929
Product: textured shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4087/330/712/2/w/448/4087330712_1_1_1.jpg?ts=1692780184926


## Modify Preprocessing with Stemming/Lemmatization

### Subtask:
Update the existing text preprocessing pipeline to include a stemming or lemmatization step (e.g., using NLTK's Porter Stemmer or WordNet Lemmatizer) after lowercasing and punctuation removal, and apply this to the 'product_name' column.


**Reasoning**:
To begin updating the text preprocessing pipeline, I need to import the `nltk` library and ensure that the necessary 'punkt' and 'wordnet' data are downloaded. This is a prerequisite for using stemming or lemmatization functionalities.



In [ ]:
import nltk
nltk.download('punkt')
nltk.download('wordnet')
print("NLTK punkt and wordnet data downloaded.")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


NLTK punkt and wordnet data downloaded.


[nltk_data] Downloading package wordnet to /root/nltk_data...


**Reasoning**:
Now that the necessary NLTK data is downloaded, I will import the `PorterStemmer` and define a function to apply stemming to the product names. This function will tokenize the text, stem each word, and then rejoin them, preparing the `product_name` column for improved TF-IDF vectorization.



In [ ]:
from nltk.stem import PorterStemmer

stemmer = PorterStemmer()

def stem_text(text):
    words = text.split()
    stemmed_words = [stemmer.stem(word) for word in words]
    return ' '.join(stemmed_words)

men_data['product_name_stemmed'] = men_data['product_name'].apply(stem_text)
men_data[['product_name', 'product_name_stemmed']].head()

,product_name,product_name_stemmed
0,technical training trousers,technic train trouser
1,basic training shorts,basic train short
2,cotton hoodie,cotton hoodi
3,jogger waist cotton trousers,jogger waist cotton trouser
4,tshirt with printed logo,tshirt with print logo


**Reasoning**:
Now that the product names have been stemmed, I will re-apply TF-IDF vectorization using the newly created `product_name_stemmed` column. This will create a new TF-IDF matrix that incorporates the stemmed terms, which is essential for the enhanced similarity search.



In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer_stemmed = TfidfVectorizer()
tfidf_matrix_stemmed = tfidf_vectorizer_stemmed.fit_transform(men_data['product_name_stemmed'])

print("Shape of TF-IDF matrix with stemmed names:", tfidf_matrix_stemmed.shape)

Shape of TF-IDF matrix with stemmed names: (1478, 487)


## Implement Updated Similarity Search Function

### Subtask:
Modify the `product_similarity_search` function to ensure that the input query also undergoes the same stemming/lemmatization preprocessing steps as the product names, allowing for consistent matching against the updated TF-IDF matrix.


**Reasoning**:
To implement the updated similarity search function, I need to define `product_similarity_search_stemmed` that preprocesses the query using the defined `stem_text` function, vectorizes it with `tfidf_vectorizer_stemmed`, calculates cosine similarity against `tfidf_matrix_stemmed`, and returns the top N similar products along with their original and stemmed names, image URLs, and similarity scores.



In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

def product_similarity_search_stemmed(query, N=5):
    # Preprocess the query: lowercase, remove punctuation, and stem
    cleaned_query = query.lower()
    cleaned_query = re.sub(r'[^\w\s]', '', cleaned_query)
    stemmed_query = stem_text(cleaned_query)

    # Vectorize the cleaned and stemmed query using the stemmed TF-IDF vectorizer
    query_vector_stemmed = tfidf_vectorizer_stemmed.transform([stemmed_query])

    # Calculate cosine similarity between the query and all stemmed product names
    similarity_scores_stemmed = cosine_similarity(query_vector_stemmed, tfidf_matrix_stemmed).flatten()

    # Get the indices of the top N most similar products
    top_n_indices_stemmed = similarity_scores_stemmed.argsort()[-N:][::-1]

    # Retrieve the top N similar products and their details
    similar_products_stemmed = men_data.iloc[top_n_indices_stemmed][['product_name_stemmed', 'product_name', 'image_url']]
    similar_products_stemmed['similarity_score'] = similarity_scores_stemmed[top_n_indices_stemmed]

    return similar_products_stemmed

print("Product similarity search function (stemmed) defined.")

Product similarity search function (stemmed) defined.


**Reasoning**:
Now that the `product_similarity_search_stemmed` function is defined, I will test it with the example query 'shirts' and display the results to demonstrate improved search relevance due to stemming.



In [ ]:
query = "shirts"
results_stemmed = product_similarity_search_stemmed(query, N=5)
print(f"Top 5 similar products for query '{query}' (with stemming):")
for index, row in results_stemmed.iterrows():
    print(f"Original Product: {row['product_name']}, Stemmed Product: {row['product_name_stemmed']}, Similarity: {row['similarity_score']:.2f}, Image URL: {row['image_url']}")

Top 5 similar products for query 'shirts' (with stemming):
Original Product: shirt, Stemmed Product: shirt, Similarity: 1.00, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4087/360/700/2/w/448/4087360700_1_1_1.jpg?ts=1689148214465
Original Product: shirt, Stemmed Product: shirt, Similarity: 1.00, Image URL: https://static.zara.net/photos///2023/I/0/2/p/3268/306/707/2/w/448/3268306707_1_1_1.jpg?ts=1692174425473
Original Product: textured shirt, Stemmed Product: textur shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4087/330/712/2/w/448/4087330712_1_1_1.jpg?ts=1692780184926
Original Product: textured shirt, Stemmed Product: textur shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/2092/300/250/2/w/448/2092300250_1_1_1.jpg?ts=1690536385550
Original Product: textured shirt, Stemmed Product: textur shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/3665/304/251/2/w/0/3665304251_1_1_1.j

**Reasoning**:
Now that the `product_similarity_search_stemmed` function is working correctly with 'shirts', I will test it with a variety of other category queries ('shorts', 'trousers', 'shoes', 'perfumes', 'jackets') to demonstrate its improved relevance across different product types as required by the task.



In [ ]:
queries = ['shorts', 'trousers', 'shoes', 'jackets', 'hoodie', 'tshirt','blazers']

for query_term in queries:
    print(f"\n--- Searching for '{query_term}' (with stemming) ---")
    results_stemmed = product_similarity_search_stemmed(query_term, N=5)
    for index, row in results_stemmed.iterrows():
        print(f"Original Product: {row['product_name']}, Stemmed Product: {row['product_name_stemmed']}, Similarity: {row['similarity_score']:.2f}, Image URL: {row['image_url']}")


--- Searching for 'shorts' (with stemming) ---
Original Product: short jacquard bermuda shorts, Stemmed Product: short jacquard bermuda short, Similarity: 0.78, Image URL: https://static.zara.net/photos///2023/I/0/2/p/5862/304/251/2/w/448/5862304251_1_1_1.jpg?ts=1691079126778
Original Product: training shorts, Stemmed Product: train short, Similarity: 0.60, Image URL: https://static.zara.net/photos///2023/I/0/1/p/0765/304/641/612/w/448/0765304641_1_1_1.jpg?ts=1688385051022
Original Product: training shorts, Stemmed Product: train short, Similarity: 0.60, Image URL: https://static.zara.net/photos///2023/I/0/1/p/0765/304/641/612/w/0/0765304641_1_1_1.jpg?ts=1688385051022
Original Product: textured bermuda shorts, Stemmed Product: textur bermuda short, Similarity: 0.58, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4840/402/409/2/w/448/4840402409_1_1_1.jpg?ts=1688658455311
Original Product: striped bermuda shorts, Stemmed Product: stripe bermuda short, Similarity: 0.54, Image U

**Reasoning**:
The previous search results showed that the query 'perfumes' still yields irrelevant results with 0.0 similarity. This suggests that the term 'perfumes' or its stemmed form might not be present in the TF-IDF vocabulary or the product names. To debug this, I need to check if 'perfum' (the stemmed form of 'perfumes') exists in the `tfidf_vectorizer_stemmed`'s vocabulary and also if any products in the `men_data` DataFrame (specifically in the `product_name_stemmed` column) contain this term.



## Summary:

### Data Analysis Key Findings

*   **Improved Search Relevance with Stemming**: Integrating stemming into the preprocessing pipeline significantly enhanced search relevance. For example, a query for 'shirts' that previously yielded 0.0 similarity now correctly identified products like 'shirt' and 'textured shirt' with high similarity scores (1.00 and 0.61, respectively). Similar improvements were observed for queries such as 'shorts', 'trousers', 'shoes', and 'jackets'.
*   **Successful Stemming Implementation**: Product names were successfully stemmed using NLTK's Porter Stemmer, creating a new `product_name_stemmed` column. For instance, "technical training trousers" became "technic train trouser," and "basic training shorts" became "basic train short." This stemmed data was then used to re-vectorize with TF-IDF, resulting in a matrix of (1478, 487) unique stemmed terms.
*   **Consistent Query Preprocessing**: The `product_similarity_search_stemmed` function was successfully implemented to apply the same stemming preprocessing to user queries, ensuring consistency between query and product name representations for accurate similarity calculations.
*   **Identified Data Gap for Specific Categories**: The search for 'perfumes' consistently yielded irrelevant results with 0.0 similarity. An investigation revealed that the stemmed term 'perfum' was not present in the TF-IDF vocabulary, nor in any of the product names within the dataset, indicating a data limitation rather than a flaw in the search logic.

### Insights or Next Steps

*   **Enrich Dataset and Vocabulary**: To address gaps like the 'perfumes' query, the dataset should be enriched with more diverse product descriptions, or alternative stemmers/lemmatizers should be explored to better handle domain-specific terminology.
*   **Explore Advanced NLP Models**: Implement advanced natural language processing models such as Word2Vec, GloVe, or BERT embeddings to capture semantic relationships between words, potentially leading to more nuanced and accurate similarity matching beyond simple term frequency and stemming.


# Task
Develop a Python function named `recommend_products` that takes a product name as input. This function will use the `product_similarity_search_stemmed` function to find and return a list of recommended product names, their similarity scores, and their image URLs.

## Implement Recommendation Function

### Subtask:
Develop a Python function, let's call it `recommend_products`, that takes a product name as input. This function will internally use the previously defined `product_similarity_search_stemmed` function to find similar products based on the input product's name. It will return a list of recommended product names, their similarity scores, and their image URLs.


**Reasoning**:
To implement the `recommend_products` function, I will define it to take a product name and an optional number of recommendations, call the `product_similarity_search_stemmed` function internally, and then convert its DataFrame output into a list of dictionaries as specified.



In [ ]:
def recommend_products(product_name_input, N=5):
    # Use the existing product_similarity_search_stemmed function
    recommended_df = product_similarity_search_stemmed(product_name_input, N=N)

    # Convert the DataFrame to a list of dictionaries
    recommendations_list = recommended_df.apply(lambda row: {
        'product_name': row['product_name'],
        'image_url': row['image_url'],
        'similarity_score': row['similarity_score']
    }, axis=1).tolist()

    return recommendations_list

print("Recommendation function 'recommend_products' defined.")

Recommendation function 'recommend_products' defined.


**Reasoning**:
Now that the `recommend_products` function is defined, I will demonstrate its functionality by calling it with an example product name and displaying the returned recommendations to fulfill the task.



In [ ]:
example_product_name = "shirt"
recommendations = recommend_products(example_product_name, N=5)

print(f"Recommendations for '{example_product_name}':")
for rec in recommendations:
    print(f"  Product: {rec['product_name']}, Similarity: {rec['similarity_score']:.2f}, Image URL: {rec['image_url']}")

Recommendations for 'shirt':
  Product: shirt, Similarity: 1.00, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4087/360/700/2/w/448/4087360700_1_1_1.jpg?ts=1689148214465
  Product: shirt, Similarity: 1.00, Image URL: https://static.zara.net/photos///2023/I/0/2/p/3268/306/707/2/w/448/3268306707_1_1_1.jpg?ts=1692174425473
  Product: textured shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4087/330/712/2/w/448/4087330712_1_1_1.jpg?ts=1692780184926
  Product: textured shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/2092/300/250/2/w/448/2092300250_1_1_1.jpg?ts=1690536385550
  Product: textured shirt, Similarity: 0.61, Image URL: https://static.zara.net/photos///2023/I/0/2/p/3665/304/251/2/w/0/3665304251_1_1_1.jpg?ts=1691486095272


## Test Recommendation Function

### Subtask:
Execute the `recommend_products` function with a specific product name from the dataset (e.g., 'shirt' or 'trousers') to demonstrate how it suggests other similar items. The output should clearly display the recommended product names, their similarity scores, and image URLs.


**Reasoning**:
To demonstrate the `recommend_products` function, I will call it with 'trousers' as the example product name and print the recommended products' details.



In [ ]:
example_product_name = "trousers"
recommendations = recommend_products(example_product_name, N=5)

print(f"Recommendations for '{example_product_name}':")
for rec in recommendations:
    print(f"  Product: {rec['product_name']}, Similarity: {rec['similarity_score']:.2f}, Image URL: {rec['image_url']}")

Recommendations for 'trousers':
  Product: suit trousers, Similarity: 0.60, Image URL: https://static.zara.net/photos///2023/I/0/2/p/5735/524/431/2/w/448/5735524431_1_1_1.jpg?ts=1689605447689
  Product: suit trousers, Similarity: 0.60, Image URL: https://static.zara.net/photos///2023/I/0/1/p/5624/243/500/2/w/448/5624243500_1_1_1.jpg?ts=1690883303116
  Product: cargo trousers, Similarity: 0.58, Image URL: https://static.zara.net/photos///2023/I/0/1/p/0108/421/709/2/w/448/0108421709_1_1_1.jpg?ts=1690442728649
  Product: striped textured trousers, Similarity: 0.53, Image URL: https://static.zara.net/photos///2023/I/0/2/p/4840/401/804/2/w/448/4840401804_1_1_1.jpg?ts=1689756834684
  Product: striped textured trousers, Similarity: 0.53, Image URL: https://static.zara.net/photos///2023/I/0/2/p/6917/402/712/2/w/448/6917402712_1_1_1.jpg?ts=1687856106736


In [ ]:
import joblib
# Save the vectorizer and the matrix
joblib.dump(tfidf_vectorizer_stemmed, 'tfidf_vectorizer.joblib')
joblib.dump(tfidf_matrix_stemmed, 'tfidf_matrix.joblib')
print('Model and matrix saved successfully!')

Model and matrix saved successfully!


## Summarize Results and Discuss Further Improvements

### Subtask:
Summarize the results of the recommendation function test, highlight its effectiveness and discussing any observations. Also, include potential next steps for refining the recommendation model's accuracy and capabilities, such as incorporating collaborative filtering or content-based features beyond product names.


## Summary:

### Q&A
The `recommend_products` function effectively generates recommendations based on product name similarity. It successfully identifies similar products and presents them with their respective similarity scores and image URLs.

The recommendation model currently relies solely on stemmed product name similarity. While effective for direct textual matches, it has limitations, such as not capturing broader semantic relationships, user preferences, or product attributes beyond the name.

### Data Analysis Key Findings
*   The `recommend_products` function was successfully developed to take a product name as input and return a list of recommended products, their similarity scores, and image URLs.
*   The function internally utilizes `product_similarity_search_stemmed` to find similar items.
*   For the input "shirt", the recommendations included the exact match "shirt" (similarity 1.00) and related items like "textured shirt" (similarity around 0.61).
*   For the input "trousers", the recommendations included "suit trousers" (similarity 0.60), "cargo trousers" (similarity 0.58), and "striped textured trousers" (similarity 0.53).
*   The output format correctly presented the product name, similarity score (formatted to two decimal places), and image URL for each recommendation.
*   The current recommendation effectiveness is limited to direct textual similarity of product names, as evidenced by the high similarity scores for close variations and the original product itself.

### Insights or Next Steps
*   To enhance recommendation quality, consider incorporating additional content-based features beyond just product names, such as product descriptions, categories, or attributes.
*   Explore integrating collaborative filtering techniques (e.g., user-item interactions like purchase history or ratings) to capture broader user preferences and provide more personalized recommendations.


# Task
Save the complete product recommendation model and its associated metadata to disk. This includes exporting the fitted `tfidf_vectorizer_stemmed` and the `tfidf_matrix_stemmed` as "tfidf_vectorizer.joblib" and "tfidf_matrix.joblib" respectively, and saving the cleaned `men_data` DataFrame (which contains the `image_url` and stemmed names) as a CSV file. After saving, verify the consistency of the system by reloading these components and performing a test query to ensure the matrix correctly maps back to the expected image URLs and product metadata. Finally, provide a high-level summary of the search engine architecture and explain how users can manually validate the relevance of the recommendations.

## Save Complete Model and Metadata

### Subtask:
Export the TF-IDF vectorizer, the feature matrix, and the cleaned 'men_data' DataFrame to disk to ensure portability and data integrity.


**Reasoning**:
I will export the trained TF-IDF vectorizer, the feature matrix, and the cleaned men_data DataFrame to disk using joblib and pandas to ensure data persistence and portability.



In [ ]:
import joblib
import os

# 1. Export the TF-IDF vectorizer
joblib.dump(tfidf_vectorizer_stemmed, 'tfidf_vectorizer.joblib')

# 2. Export the feature matrix
joblib.dump(tfidf_matrix_stemmed, 'tfidf_matrix.joblib')

# 3. Save the cleaned men_data DataFrame
men_data.to_csv('cleaned_men_data.csv', index=False)

# 4. Verify the files exist
files = ['tfidf_vectorizer.joblib', 'tfidf_matrix.joblib', 'cleaned_men_data.csv']
for file in files:
    if os.path.exists(file):
        print(f'Successfully saved: {file}')
    else:
        print(f'Failed to save: {file}')

# List all files in the current directory
!ls -l *.joblib *.csv

Successfully saved: tfidf_vectorizer.joblib
Successfully saved: tfidf_matrix.joblib
Successfully saved: cleaned_men_data.csv
-rw-r--r-- 1 root root 1045318 Mar  8 13:00 cleaned_men_data.csv
-rw-r--r-- 1 root root   69331 Mar  8 13:00 tfidf_matrix.joblib
-rw-r--r-- 1 root root   10125 Mar  8 13:00 tfidf_vectorizer.joblib


## Verify Recommendation Consistency

### Subtask:
Reload the saved model components and perform a test query to ensure that the reloaded system produces consistent results and correctly maps to product metadata.


**Reasoning**:
I will reload the saved model components (vectorizer, matrix, and dataframe) and implement a test to verify that the recommendation system produces consistent results using the reloaded files.



In [ ]:
import joblib
import pandas as pd
import re
from sklearn.metrics.pairwise import cosine_similarity

# 1. Load the saved components
loaded_vectorizer = joblib.load('tfidf_vectorizer.joblib')
loaded_matrix = joblib.load('tfidf_matrix.joblib')
loaded_men_data = pd.read_csv('cleaned_men_data.csv')

# 2. Define the search logic using loaded components
def verify_similarity_search(query, N=5):
    # Preprocess query exactly as before
    from nltk.stem import PorterStemmer
    stemmer = PorterStemmer()

    cleaned_query = query.lower()
    cleaned_query = re.sub(r'[^\w\s]', '', cleaned_query)
    stemmed_query = ' '.join([stemmer.stem(word) for word in cleaned_query.split()])

    # Vectorize and calculate similarity
    query_vec = loaded_vectorizer.transform([stemmed_query])
    scores = cosine_similarity(query_vec, loaded_matrix).flatten()

    top_indices = scores.argsort()[-N:][::-1]
    results = loaded_men_data.iloc[top_indices].copy()
    results['similarity_score'] = scores[top_indices]

    return results[['product_name', 'similarity_score', 'image_url']]

# 3. Perform test query
test_query = 'trousers'
verification_results = verify_similarity_search(test_query)

print(f'Verification results for query: {test_query}')
print(verification_results.to_string(index=False))

Verification results for query: trousers
             product_name  similarity_score                                                                                                image_url
            suit trousers          0.601450 https://static.zara.net/photos///2023/I/0/2/p/5735/524/431/2/w/448/5735524431_1_1_1.jpg?ts=1689605447689
            suit trousers          0.601450 https://static.zara.net/photos///2023/I/0/1/p/5624/243/500/2/w/448/5624243500_1_1_1.jpg?ts=1690883303116
           cargo trousers          0.582051 https://static.zara.net/photos///2023/I/0/1/p/0108/421/709/2/w/448/0108421709_1_1_1.jpg?ts=1690442728649
striped textured trousers          0.533098 https://static.zara.net/photos///2023/I/0/2/p/4840/401/804/2/w/448/4840401804_1_1_1.jpg?ts=1689756834684
striped textured trousers          0.533098 https://static.zara.net/photos///2023/I/0/2/p/6917/402/712/2/w/448/6917402712_1_1_1.jpg?ts=1687856106736


In [ ]:
import joblib
import pandas as pd
import re
from sklearn.metrics.pairwise import cosine_similarity
from nltk.stem import PorterStemmer

# 1. Load the saved components from disk
vectorizer = joblib.load('tfidf_vectorizer.joblib')
matrix = joblib.load('tfidf_matrix.joblib')
data = pd.read_csv('cleaned_men_data.csv')
stemmer = PorterStemmer()

def get_recommendations(query, N=5):
    # Preprocess
    clean = re.sub(r'[^\w\s]', '', query.lower())
    stemmed = ' '.join([stemmer.stem(w) for w in clean.split()])

    # Vectorize & Search
    query_vec = vectorizer.transform([stemmed])
    scores = cosine_similarity(query_vec, matrix).flatten()

    # Get Top N
    top_idx = scores.argsort()[-N:][::-1]
    results = data.iloc[top_idx].copy()
    results['score'] = scores[top_idx]
    return results[['product_name', 'score', 'image_url']]

# Test it with a category
my_request = 'jackets'
final_results = get_recommendations(my_request)

print(f'Recommendations for: {my_request}')
display(final_results)

Recommendations for: jackets


,product_name,score,image_url
536,textured jacket,0.775687,https://static.zara.net/photos///2023/I/0/2/p/...
574,cotton jacket,0.730259,https://static.zara.net/photos///2023/I/0/2/p/...
568,technical jacket,0.689525,https://static.zara.net/photos///2023/I/0/2/p/...
538,technical jacket,0.689525,https://static.zara.net/photos///2023/I/0/2/p/...
539,contrast print jacket,0.652782,https://static.zara.net/photos///2023/I/0/2/p/...
